### PR TITLE
Rewrite Soft NMS with pytorch extension

### PR DESCRIPTION
**Pull request by xvjiarui**
1. Replaced Cython soft NMS with PyTorch extension
2. Remove Cython dependency.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -261,6 +235,10 @@ def gen_packages_items():
                 name='nms_cuda',
                 module='mmdet.ops.nms',
                 sources=['src/nms_cuda.cpp', 'src/nms_kernel.cu']),
+            make_cuda_ext(
+                name='soft_nms_cpu',
+                module='mmdet.ops.nms',
+                sources=['src/soft_nms_cpu.cpp']),
```
Comment (hellock) [inline setup.py]: I suggest to use only one extension for NMS and SoftNMS.

Comment (xvjiarui) [reply setup.py]: I will merge them into one cpp file.

File: mmdet/ops/nms/src/nms_cpu.cpp [hunk 2]
```diff
@@ -66,6 +66,149 @@ at::Tensor nms(const at::Tensor& dets, const float threshold) {
   return result;
 }
 
+template <typename scalar_t>
```
Comment (hellock) [inline mmdet/ops/nms/src/nms_cpu.cpp]: We can add a line "Modified from https://xxxx, Soft-NMS is added." at the top of this file.

File: mmdet/ops/nms/nms_wrapper.py [hunk 3]
```diff
@@ -76,10 +75,10 @@ def soft_nms(dets, iou_thr, method='linear', sigma=0.5, min_score=1e-3):
     """
```
Comment (hellock) [inline mmdet/ops/nms/nms_wrapper.py]: We may add a standard docstring for this method like in `nms()`.

```python
def soft_nms(arg1, arg2=0.5):
    """A brief description.

    Detailed description.

    Args:
        <arg1> (<type>): <Description>.
        <arg2> (<arg2>): <Description>. <Default: 0.5>.

    Returns:
        <type>: <Description>.
    """
    pass
```

File: tests/test_soft_nms.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    assert len(inds) == len(new_dets) == 4
+
+    dets = torch.FloatTensor(base_dets)
+    new_dets, inds = soft_nms(dets, iou_thr)
+    assert dets.dtype == new_dets.dtype
+    assert len(inds) == len(new_dets) == 4
+
+    dets = base_dets.astype(np.float64)
+    new_dets, inds = soft_nms(dets, iou_thr)
+    assert dets.dtype == new_dets.dtype, "{} vs {}".format(
```
Comment (hellock) [inline tests/test_soft_nms.py]: Single quote is preferred in this codebase.

**Pull request merged**